systemdev/testing-locally.rst

```diff
@@ -3,18 +3,207 @@ Making changes and testing locally
 
 On this page you'll find information on how to build Ubuntu Touch system software for your device. Most of the software preinstalled on your Ubuntu Touch device is shipped in the device image in the form of a Debian package. This format is used by several Linux distributions, such as Debian, Ubuntu, and Linux Mint. Plenty of `documentation on deb packages <https://www.debian.org/doc/manuals/maint-guide/index.en.html>`__ is available, so we won't be covering it here. Besides, in most cases you'll find yourself in need of modifying existing software rather than developing new packages from scratch. For this reason, this guide is mostly about recompiling an existing Ubuntu Touch package.
 
-There are essentially two ways of developing Ubuntu Touch system software locally:
+There are essentially three ways of developing Ubuntu Touch system software locally:
 
+* `Building packages in a chroot using sbuild`_
 * `Cross-building with crossbuilder`_
 * `Building on the device itself`_
 
-We'll examine both methods, using `address-book-app <https://github.com/ubports/address-book-app>`__ (the Contacts application) as an example.
+sbuild uses a more minimal chroot-based build environment than the LXD container used by crossbuilder which might miss certain build dependencies if they are pre-installed in the LXD image.  In contrast to crossbuilder, sbuild automatically creates a log file and runs `lintian <https://lintian.debian.org/manual/lintian.html>`__ on the built packages in order to detect any problems.  Due to the creation of a persistent LXD container per package, crossbuilder can be quicker for subsequent builds of the same package than sbuild which installs all build dependencies on each run.  The use of LXD also allows for easier inspection, debugging and manual modification of the build environment.  crossbuilder can also automatically deploy build packages on a connected device via ADB.
+
+We'll examine the use of crossbuilder and builds on the device using `address-book-app <https://github.com/ubports/address-book-app>`__ (the Contacts application) as an example.
 
 We only recommend developing packages using a device with Ubuntu Touch installed from the devel channel. This ensures that you are testing your changes against the most current state of the Ubuntu Touch code.
 
-.. warning::
+Preparing a package for a build
+-------------------------------
 
-    Installing packages has a risk of damaging the software on your device, rendering it unusable. If this happens, you can :doc:`reinstall Ubuntu Touch </userguide/install>`.
+``sbuild`` or ``crossbuilder`` need to be invoked from a debianized package source tree (i.e. the package sources with a ``debian`` subdirectory), UBports packages mostly consists of git repositories containing native or non-native packages.  Native packages can be built directly and the following script can be used in order to prepare a non-native package for a build::
```

Comment:
Debian package formats are well defined and more or less technically specified in [dpkg-source(1)](https://manpages.debian.org/unstable/dpkg-dev/dpkg-source.1.en.html#SOURCE_PACKAGE_FORMATS).

In UBports native packages either format 1.0 or 3.0 consist of a git repo with the full sources and modifications as well as the debian directory. These are typically packages for which the UBports project is the upstream, an example would be [lomiri-thumbnailer](https://gitlab.com/ubports/core/lomiri-thumbnailer/).

Non-native packages in format 3.0 (quilt) separates debian packaging, patches, and upstream sources. In UBports the ./debian/ubport.source_location usually contains the download location of the upstream tarball. It is the preferred format for packages for which UBports is not the upstream since it keeps downstream modifications separate. Since such a git repo only contains debian directory with patches, it needs some preparation. That is the source tarball needs to be fetched, extracted and debianized. An example would be [libgbinder-radio](https://gitlab.com/ubports/core/packaging/libgbinder-radio).